File: opensource/kernel/src/main/java/com/sos/scheduler/engine/main/SchedulerController.java
package com.sos.scheduler.engine.main;

import com.sos.scheduler.engine.common.time.Time;
import com.sos.scheduler.engine.eventbus.SchedulerEventBus;

/** Steuerung für den C++-Scheduler in einem eigenen nebenläufigen Thread. */
public interface SchedulerController {
    /** Startet den Scheduler ohne zu warten. */
    void startScheduler(Iterable<String> arguments);

    /** Veranlasst die Beendigung des Schedulers, wartet aufs Ende und schließt alles. */
    void close();

    /** Veranlasst die Beendigung des Schedulers. */
    void terminateScheduler();

    /** Veranlasst die Beendigung des Schedulers nach einem Fehler.
     * Kann aus einem anderen Thread aufgerufen werden und lässt die Warte-Methoden die Exception werfen. */
    void terminateAfterException(Throwable x);

    boolean tryWaitForTermination(Time timeout);

    int exitCode();

    SchedulerEventBus getEventBus();
}

File: engine-kernel/src/main/scala/com/sos/scheduler/engine/main/SchedulerController.scala
package com.sos.scheduler.engine.main

import com.google.inject.Injector
import com.sos.scheduler.engine.eventbus.SchedulerEventBus
import java.time.Duration

/**
 * Steuerung für den C++-Scheduler in einem eigenen nebenläufigen Thread.
 * @author Joacim Zschimmer
 */
trait SchedulerController extends AutoCloseable {
  def startScheduler(arguments: String*): Unit

  /** Veranlasst die Beendigung des Schedulers, wartet aufs Ende und schließt alles. */
  def close(): Unit

  /** Veranlasst die Beendigung des Schedulers. */
  def terminateScheduler(): Unit

  /** Veranlasst die Beendigung des Schedulers nach einem Fehler.
   * Kann aus einem anderen Thread aufgerufen werden und lässt die Warte-Methoden die Exception werfen. */
  def terminateAfterException(x: Throwable): Unit

  def tryWaitForTermination(timeout: Duration): Boolean

  def exitCode: Int

  def eventBus: SchedulerEventBus

  def injector: Injector
}
